---
title: "analyze_evaluation_results.Rmd"
author: "niko.popitsch@imba.oeaw.ac.at"
documentclass: article
fontsize: 10pt
output:
  html_document:
    df_print: paged
  pdf_document:
    fig_caption: yes
    fig_height: 10
    fig_width: 10
geometry: margin=1in
classoption: a4paper
params:
    extra_transcript_table:
       value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/data/slamstr/pooled.mm10.all/pooled.mm10.all.Slamstr.stats.trans.final.tsv.gz"  
    # splice_sim_config:
    #    value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/big3_biseq_nf/splice_sim.config.json"
    # out_dir:
    #    value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/big3_biseq_nf/analysis/"
    # plot_type:
    #    value: "by_mapper"
    # splice_sim_config:
    #    value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/big3_slamseq_nf/splice_sim.config.json"
    # out_dir:
    #     value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/big3_slamseq_nf/analysis/"
    splice_sim_config:
       value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/big4_slamseq_nf/splice_sim.config.json"
    out_dir:
       value: "/Volumes/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/big4_slamseq_nf/analysis/"
    plot_type:
      value: "by_condition"

---
<style type="text/css">
body, td {
   font-size: 10px;
}
code.r{
  font-size: 10px;
}
pre {
  font-size: 10px
}
</style>

# INIT

```{r setup, include=FALSE}
require(data.table)
require(tidyr)
require(dplyr)
require(ggplot2)
require(scales)
require(rjson)
require(stringr)
require(VGAM)
require(cowplot)
require(arrow)
require(tictoc)
require(ggpubr)

# to ensure stripping '\0' (nul) from character vector
options(arrow.skip_nul = TRUE)

# load data from TSV
load_table = function(dataF, append="", header=T) {
  dataF = as.character(paste0(dataF, append))
  print(paste("Loading", dataF))
  if ( endsWith(dataF, ".gz") ) {
    return(fread(cmd=paste('gunzip -c', dataF), header=header, sep="\t", na.strings=c("na","NA",".", "None")))
  } else {
    return(fread(dataF, header=header, sep="\t", na.strings=c("na","NA",".", "None")))
  }
}

# multiple plots with single title
my_plot_grid = function(plots, main, ncol=NULL, nrow=NULL) {
  plot_row=plot_grid(plotlist=plots, ncol=ncol, nrow=nrow)
  title <- ggdraw() + draw_label( main, fontface = 'bold', x = 0, hjust = 0 ) + theme(plot.margin = margin(0, 0, 0, 7))
  return (plot_grid(title, plot_row, ncol = 1, rel_heights = c(0.1, 1)))
}

# calculate confidence interval
# call with calc_ci( mtcars, mpg)
calc_ci = function(d, col, min_lower_ci=NA, max_upper_ci=NA) {
  ret = d %>% summarise(
            col.mean := mean({{col}}, na.rm = TRUE),
            col.sd = sd({{col}}, na.rm = TRUE),
            col.n = n()) %>%
  mutate(stderr = col.sd / sqrt(col.n),
         lower.ci = col.mean - qt(1 - (0.05 / 2), col.n - 1) * stderr,
         upper.ci = col.mean + qt(1 - (0.05 / 2), col.n - 1) * stderr) %>% 
  ungroup() %>% 
  mutate(lower.ci = pmax(min_lower_ci, lower.ci, na.rm=T),
         upper.ci = pmin(max_upper_ci, upper.ci, na.rm=T))
  return (ret)
}
```

# data
```{r data, include=FALSE}

# # load evaluation results
# data_file=paste0(params$out_dir,'/evaluation_results.rds')
# if (file.exists(data_file) ) {
#   d = readRDS(data_file)
# } else {
#   # additional transcript md
#   gene_anno=load_table(paste0(home_dir, '/sim/reference_model/gene_anno.tsv.gz')) %>% mutate(len=end-start+1)
#   extra_transcript_table=load_table(params$extra_transcript_table) %>% select(transcript_id, rnk, GC_frac, mappability_median, mappability_mean, T_Pos)
#   # load data per dataset
#   tid_performance=tibble()
#   reads=tibble()
#   splice_site_performance=tibble()
#   splice_site_mappability=tibble()
#   feature_overlap=tibble()
#   transcript_info=tibble()
#   conversion_rates=c('0', as.character(conf$condition$conversion_rates))
#   for ( m in c( 'truth', names(conf$mappers))) {
#     for ( cr in conversion_rates) {
#       pre=paste0(home_dir, '/eva/overall_performance/', conf$dataset_name, '.cr',cr, '.', m)
#       tid_performance=tid_performance %>% rbind(load_table(paste0(pre, '.tid_performance.tsv')))
#       if (file.exists(paste0(pre, '.reads.tsv'))) {
#         reads=reads %>% rbind(load_table(paste0(pre, '.reads.tsv')))
#       }
#       pre=paste0(home_dir, '/eva/splice_site_performance/', conf$dataset_name, '.cr',cr, '.', m)
#       if (file.exists(paste0(pre, '.splice_site_performance.tsv'))) {
#         splice_site_performance=splice_site_performance %>% rbind(load_table(paste0(pre, '.splice_site_performance.tsv')))
#       }
#       pre=paste0(home_dir, '/eva/splice_site_mappability/', conf$dataset_name, '.cr',cr, '.', m)
#       if (file.exists(paste0(pre, '.mappability.tsv'))) {
#         splice_site_mappability=splice_site_mappability %>% rbind(load_table(paste0(pre, '.mappability.tsv')))
#       }
#       pre=paste0(home_dir, '/eva/feature_overlap/', conf$dataset_name, '.cr',cr, '.', m)
#       #if (file.exists(paste0(pre, '.feature_counts.tsv'))) {
#       #  feature_overlap=feature_overlap %>% rbind(load_table(paste0(pre, '.feature_counts.tsv')))
#       #}
#       if (file.exists(paste0(pre, '.transcript_info.tsv'))) {
#         transcript_info=transcript_info %>% rbind(load_table(paste0(pre, '.transcript_info.tsv')))
#       }
#     }
#   }
#   
#   # add mappability, coords, etc.
#   tid_performance=tid_performance %>% 
#     merge(gene_anno, by.x='tid', by.y='transcript_id', all.x=T) %>% 
#     merge(extra_transcript_table, by.x='tid', by.y='transcript_id', all.x=T) %>% 
#     mutate(mapper=factor(mapper),
#            condition_id=factor(condition_id),
#            precision=TP/(TP+FP), 
#            recall=TP/(TP+FN), 
#            F1=2*TP/(2*TP+FP+FN)
#     )
#   # combine
#   d=list()
#   d[['tid_performance']]=tid_performance
#   d[['reads']]=reads
#   d[['splice_site_performance']]=splice_site_performance
#   d[['splice_site_mappability']]=splice_site_mappability
#   d[['gene_anno']]=gene_anno
#   d[['extra_transcript_table']]=extra_transcript_table
#   d[['feature_overlap']]=feature_overlap
#   d[['transcript_info']]=transcript_info
#   saveRDS(d, data_file)
# }

home_dir=paste0(dirname(params$splice_sim_config),'/')
conf=fromJSON(paste(readLines(params$splice_sim_config), collapse=""))

tic("load evaluation data") # ca. 5-10min.
d=list()
d[['tid_performance']]=open_dataset(paste0(home_dir,'/results/overall_performance/')) %>% collect()
d[['splice_site_performance']]=open_dataset(paste0(home_dir,'/results/splice_site_performance/')) %>% collect()
d[['splice_site_mappability']]=open_dataset(paste0(home_dir,'/results/splice_site_mappability/')) %>% collect()
d[['gene_anno']]=open_dataset(paste0(home_dir,'/results/gene_anno/')) %>% collect()
fc_schema=schema(
  condition_id = string(), # need to set explicit types as otherwise arrow thinks this is an int
  tid = string(),
  fid = string(),
  start = int64(),
  end = int64(),
  FP = int64(),
  FN = int64(),
  mapper = string(),
  chromosome = string(),
  ftype = string()
)
d[['feature_counts']]=open_dataset(paste0(home_dir,'/results/feature_counts/'), schema = fc_schema) # too big, do not collect
d[['transcript_info']]=open_dataset(paste0(home_dir,'/results/transcript_info/')) %>% collect()
# optional: read info
if (file.exists(paste0(home_dir,'/results/reads/')) ) {
  d[['reads']]=open_dataset(paste0(home_dir,'/results/reads/')) # too big, do not collect
} else {
  d[['reads']]=NA
}
d[['extra_transcript_table']]=load_table(params$extra_transcript_table) %>% select(transcript_id, rnk, GC_frac, mappability_median, mappability_mean, T_Pos)
conversion_rates=c('0', as.character(conf$condition$conversion_rates))
toc()

tic("data cleaning") 
# data cleaning ~1min
# add mappability, coords, etc.
d[['gene_anno']] = d[['gene_anno']] %>% 
  mutate(len=end-start+1) %>% 
  rename(tid=transcript_id)
d[['extra_transcript_table']] = d[['extra_transcript_table']] %>% 
  rename(tid=transcript_id) %>% 
  select(-rnk)
d[['tid_performance']]=d[['tid_performance']] %>% 
  merge(d[['gene_anno']], by='tid', all.x=T) %>% 
  merge(d[['extra_transcript_table']], by='tid', all.x=T) %>% 
  mutate(mapper=factor(mapper),
         condition_id=factor(condition_id),
         precision=TP/(TP+FP), 
         recall=TP/(TP+FN), 
         F1=2*TP/(2*TP+FP+FN)
  )
d[['splice_site_mappability']] = d[['splice_site_mappability']] %>% 
  mutate(
         acc_win_min_map=replace(acc_win_min_map, acc_win_min_map == "None", 0), # set mappability to zero if 'None' reported
         acc_win_max_map=replace(acc_win_max_map, acc_win_max_map == "None", 0),
         don_win_min_map=replace(don_win_min_map, don_win_min_map == "None", 0),
         don_win_max_map=replace(don_win_max_map, don_win_max_map == "None", 0),
  ) %>% 
  mutate(acc_win_min_map=as.numeric(acc_win_min_map),
         acc_win_max_map=as.numeric(acc_win_max_map),
         don_win_min_map=as.numeric(don_win_min_map),
         don_win_max_map=as.numeric(don_win_max_map))
d[['splice_site_performance']] = d[['splice_site_performance']] %>% 
  merge(d[['gene_anno']], by='tid', all.x=T) %>% 
  merge(d[['extra_transcript_table']], by='tid', all.x=T) %>% 
  mutate(mapper=factor(mapper),
         condition_id=factor(condition),
         spl_pre=spl_TP/(spl_TP+spl_FP), 
         spl_rec=spl_TP/(spl_TP+spl_FN), 
         spl_F1=2*spl_TP/(2*spl_TP+spl_FP+spl_FN),
         don_pre=don_TP/(don_TP+don_FP), 
         don_rec=don_TP/(don_TP+don_FN), 
         don_F1=2*don_TP/(2*don_TP+don_FP+don_FN),
         acc_pre=acc_TP/(acc_TP+acc_FP), 
         acc_rec=acc_TP/(acc_TP+acc_FN), 
         acc_F1=2*acc_TP/(2*acc_TP+acc_FP+acc_FN)
  )
toc()

```

# QC

## Data completeness
```{r}

# count non-NA values per column for all passed tables
count_na_per_column = function(d, tables) {
 counts=tibble()
  for ( tab in tables ) {
    nr=nrow(d[[tab]])
    if (nr>0) {
      counts=counts %>% rbind(
        d[[tab]] %>% select(everything()) %>% summarise_all(funs(sum(is.na(.)))) %>% mutate(table=!!tab) %>% collect() %>% pivot_longer(-table) %>% mutate(nr=!!nr, frac=value/nr)
      )
    }
  }
  return (counts)
}

# count non NA values per column in all tables (long!)
#non_na_counts = count_na_per_column(d, names(d))
non_na_counts = count_na_per_column(d, c("tid_performance","splice_site_performance","splice_site_mappability","gene_anno","transcript_info","extra_transcript_table"))
ggplot(non_na_counts, aes(x=name, y=frac)) +
  geom_bar(stat='identity') +
  facet_wrap(table~., scales = 'free') +
  ggtitle("Fraction of NA counts per table column") + 
  xlab("") + ylab("") +
  ylim(0,1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggsave(paste0(params$out_dir,'/qc_na_counts.pdf'), width = 10, height = 8)
```
## Summary stats
```{r}
pdf(paste0(params$out_dir,'/qc_summary_stats.pdf'), width = 10, height = 16)
d[['splice_site_mappability']] %>% 
  group_by(mapper) %>% slice_sample(n=10000) %>% ungroup() %>% 
  select(-c(is_converted_bam,condition,tid,chromosome,start,end,strand )) %>% 
  pivot_longer(-c('intron_id','mapper')) %>% 
  ggsummarystats(x = "name", y = "value", color="name",
    ggfunc = ggboxplot, add = NA,
    palette = "npg",
    facet.by = c("mapper")) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggtitle("Summary stats for splice_site_mappability (subsample, n=10k)")

dev.off()
```

## Missing tids

Some of the originally configured transcript ids are missing from result data. 
This is either because 
- there were no reads simulated as the annotation is smaller than read-size and there are no (longer) overlapping transcripts
- annotations are very long (e.g., Kcnip1 or Cmss1)

```{r}

all_tids=load_table(paste0(home_dir, 'tids.tsv')) %>% pull(transcript_id)
found_tids=unique(d[['tid_performance']] %>% pull(tid))
print(paste0("We have data for ", length(found_tids), '/', length(all_tids), " transcripts"))

# what tids are missing?
extra_transcript_table2=load_table(params$extra_transcript_table) %>% 
  filter(transcript_id %in% all_tids) %>% 
  select(transcript_id, chr,start,end,strand,gene_name,len,rnk, GC_frac, mappability_median, mappability_mean, T_Pos) %>% 
  mutate(missing=ifelse(transcript_id %in% found_tids, 0,1))
extra_transcript_table2 %>% ggplot(aes(x=factor(missing), y=len)) + 
  geom_violin()  +scale_y_log10() + ggtitle("Length distribution of missing transcripts")
ggsave(paste0(params$out_dir,'/qc_missing_transcripts.pdf'), width = 10, height = 8)

# did we recover all reads?
d[['tid_performance']] %>% mutate(all=TP+FN) %>% group_by(mapper, condition_id) %>% summarise(tc=sum(all))

```

## transcript overlaps
```{r}

d[['transcript_info']] %>% count(n_overlapping) %>% 
  ggplot(aes(x=n_overlapping, y=n)) + 
  geom_bar(stat='identity') +
  ggtitle("Histogram of number of overlapping annotations per transcript")
# example: Sfmbt2 is a miRNA host gene with 120 overlapping annotations
```
## Compare numbers for unspliced transcripts
```{r}
tab = d[['tid_performance']] %>% 
  filter(rnk==1, condition_id=='0.1') %>% 
  select(mapper, tid, iso, TP, FP, FN, F1) %>% 
  pivot_wider(names_from = iso, values_from = c(TP,FP,FN,F1), id_cols = c('tid', 'mapper')) %>%
  mutate(f1_diff=F1_mat-F1_pre) %>% 
  merge(d[['gene_anno']], by = 'tid') %>% 
  merge(d[['extra_transcript_table']], by = 'tid')

p1 = tab %>% ggplot(aes(x=f1_diff)) + geom_density() + ggtitle("Difference in F1 value between pre/mat for unspliced transcripts. Should be zero!") +
  facet_wrap(mapper~.)
p2 = tab %>% filter(mappability_mean>0.5) %>% ggplot(aes(x=f1_diff)) + geom_density() + ggtitle("Difference in F1 value between pre/mat for unspliced transcripts. Map>0.5. Should be zero!") +
  facet_wrap(mapper~.)
p3 = tab %>% ggplot(aes(x=FP_mat, y=FP_pre, col=mappability_mean, alpha=0.1)) +
  geom_point() +
  scale_x_sqrt() +
  scale_y_sqrt() +
  facet_wrap(mapper~.)
p4 = tab %>% ggplot(aes(x=FN_mat, y=FN_pre, col=mappability_mean, alpha=0.1)) +
  geom_point() +
  scale_x_sqrt() +
  scale_y_sqrt() +
  facet_wrap(mapper~.)

my_plot_grid(list(p1,p2,p3,p4), main="QC: comparing isoform performance of unspliced transcripts (cr=0.1)") 
ggsave(paste0(params$out_dir,'/qc_isoform_comparison_of_unspliced_transcripts.pdf'), width = 10, height = 8)

```

# Performance 

## Performance per transcript
```{r performance}
plot_median_performance_by_bin = function(tab, subtitle='', facet=NA) {
  group_vars=c('mapper', 'condition_id', 'iso', 'is_converted_bam')
  if (!is.na(facet)) {
    # rename by addint '(n=xxx)')
    cnts = tab %>% 
      group_by_at(facet) %>% 
      summarise(n_tid=n_distinct(tid)) %>% 
      setNames(., c("a", "b")) %>% mutate(c=paste0(a,' (n=',b,')'))
    labs = setNames(cnts$c, cnts$a)
    tab[['facet']]=factor( labs[tab[[facet]]], levels=cnts$c)
    group_vars=c(group_vars, 'facet')
  }
  dat = tab %>% group_by_at(group_vars) %>% 
    summarise(median_F1=median(F1, na.rm = T))
  if ( params$plot_type=='by_condition') {
    p=dat %>% ggplot(aes(x=condition_id, y=median_F1, col=mapper, linetype=iso, group=paste0(mapper, iso))) + 
      geom_line() + 
      geom_point() +
      #facet_wrap(is_converted_bam~.) +
      ggtitle(paste0("Median F1 measure per transcript (n=",length(unique(tab$tid)),")"), subtitle)
    if (!is.na(facet)) {
      p=p+facet_wrap(facet~.)
    }
  } else {
    p=dat %>% ggplot(aes(x=mapper, y=median_F1, col=mapper, shape=iso, group=paste0(mapper, iso))) + 
      geom_point() +
      #facet_wrap(is_converted_bam~.) +
      ggtitle(paste0("Median F1 measure per transcript (n=",length(unique(tab$tid)),")"), subtitle) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    if (!is.na(facet)) {
      p=p+facet_wrap(facet~.)
    }
  }
  return(p)
}

tab = d[['tid_performance']] %>% 
  mutate(num_exons=case_when(rnk<=5 ~ as.character(rnk), 
                             rnk>5 ~ '>5',
                             T ~ 'NA'),
         mappability=case_when(mappability_mean<0.2 ~ 'low',
                               mappability_mean>0.9 ~ 'high',
                               T ~ 'medium')
  ) %>% mutate(
    num_exons=factor(num_exons, levels = c('1','2','3','4','5','>5')),
    mappability=factor(mappability, levels=c('high', 'medium', 'low'))
  )
p1 = plot_median_performance_by_bin(tab, 'by conversion rate', facet=NA)
p2 = plot_median_performance_by_bin(tab, 'by regular read mappability', facet='mappability')
p3 = plot_median_performance_by_bin(tab %>% filter(mappability=='high'), 'by number of exons (high mappability only)', facet='num_exons')
p4 = plot_median_performance_by_bin(tab %>% filter(mappability=='medium'), 'by number of exons (medium mappability only)', facet='num_exons')
#p4 = plot_median_performance_by_bin(tab %>% filter(mappability=='high'), 'by transcript_type (high mappability only)', facet='transcript_type')

my_plot_grid(list(p1,p2,p3,p4), main="Overall mapping performance per transcript") 
ggsave(paste0(params$out_dir,'/tid_performance.pdf'), width = 10, height = 8)

```

## Performance correlations
```{r correlations}

# some correlations
plot_corr = function(d, a, b, subtitle='', xlog=F, draw_diag=T) {
  thecor = paste("r_pearson = ", round(cor(d[[a]], d[[b]], use = "complete.obs"), 4), 
                 "\nr_spearman = ", round(cor(d[[a]], d[[b]], use = "complete.obs", method="spearman"), 4),
                 "\nn =",nrow(d)  )
  p = ggplot( d %>% filter(!is.na(mapper)), aes_string(x=a, y=b) ) +
    #geom_point(aes(col=condition_id, alpha=0.5), size=.5)  +
    #annotation_custom(thecor) + 
    #geom_point(aes(col=iso, alpha=0.05), size=.5)  +
    stat_binhex(bins=100) + scale_fill_gradientn("", colours = rev(rainbow(5, end = 4/6))) +
    facet_grid(iso ~ mapper) +
    ggtitle(paste0("Correlation between ",a," and ",b, '\n', subtitle), thecor)
  if (draw_diag) {
    p=p+geom_abline(intercept = 0, slope = 1, col="black",linetype="dotted") 
  }
  if ( xlog ) {
    p=p+scale_x_log10()
  }
  return(p)
}

tab = d[['tid_performance']] %>% 
  mutate(num_exons=rnk,
         mappability=case_when(mappability_mean<0.2 ~ 'low',
                               mappability_mean>0.9 ~ 'high',
                               T ~ 'medium')
  ) %>% mutate(
    mappability=factor(mappability, levels=c('high', 'medium', 'low'))
  )

p1=plot_corr(tab %>% filter(mapper != 'truth', condition_id=='0.1'), "mappability_mean", "F1")
p2=plot_corr(tab %>% filter(mapper != 'truth', condition_id=='0.1'), "GC_frac", "F1")  
p3=plot_corr(tab %>% filter(mapper != 'truth', condition_id=='0.1'), "len", "F1", xlog=T, draw_diag=F)  
p4=plot_corr(tab %>% filter(mapper != 'truth', condition_id=='0.1'), "num_exons", "F1", draw_diag=F)  
p5=plot_corr(tab %>% filter(mapper != 'truth', condition_id=='0.1', mappability=='high'), "num_exons", "F1", draw_diag=F, subtitle = 'high mappability only')  
p6=plot_corr(tab %>% filter(mapper != 'truth', mappability=='medium'), "num_exons", "F1", draw_diag=F, subtitle = 'medium mappability only')  
my_plot_grid(list(p1,p2,p3,p4,p5,p6), main="Correlations with mapping performance") 
ggsave(paste0(params$out_dir,'/tid_performance_corr.pdf'), width = 16, height = 12)

```

## mismapped reads
```{r mismapped_reads}

plot_mismapped_reads = function(d, subtitle='') {
  dat = d %>% group_by(mapper, condition_id, iso, is_converted_bam) %>% 
    summarise(median_F1=median(F1, na.rm = T))
  if ( params$plot_type=='by_condition' ) {
    p=dat %>% ggplot(aes(x=condition_id, y=median_F1, col=mapper, linetype=iso, group=paste0(mapper, iso))) + 
      geom_line() + geom_point() +
      facet_wrap(is_converted_bam~.) +
      ggtitle(paste0("Median F1 measure per transcript (n=",length(unique(d$tid)),")"), subtitle)
  } else {
    p=dat %>% ggplot(aes(x=mapper, y=median_F1, col=mapper, shape=iso, group=paste0(mapper, iso))) + 
      geom_point() +
      facet_wrap(is_converted_bam~.) +
      ggtitle(paste0("Median F1 measure per transcript (n=",length(unique(d$tid)),")"), subtitle) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  }
  return(p)
}


tab = reads %>% group_by(classification, is_converted_bam, mapper, condition_id, true_isoform) %>% count() 
if (params$plot_type=='by_condition') {
  p1 = tab %>% ggplot(aes(x=condition_id, y=n, col=mapper, linetype=true_isoform, group=paste0(mapper, true_isoform, classification))) + 
    geom_line() + 
    facet_wrap(classification~.) +
    ggtitle( "Mismapped reads per mapper" )
} else {
  p1 = tab %>% ggplot(aes(x=mapper, y=n, col=mapper, shape=true_isoform, group=paste0(mapper, true_isoform, classification))) + 
    geom_point() + 
    facet_wrap(classification~.) +
    ggtitle( "Mismapped reads per mapper" ) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}

tab = reads %>% filter(is_converted_bam==1) %>% 
  group_by(classification, mapper, condition_id, true_isoform) %>% summarise(m_ntc=mean(n_tc_pos, na.rm=T))
if (params$plot_type=='by_condition') {
  p2 = tab %>% ggplot(aes(x=condition_id, y=m_ntc, col=mapper, linetype=true_isoform, group=paste0(mapper, true_isoform, classification))) + 
    geom_line() +
    facet_wrap(classification~.) +
    ggtitle( "Mean n_tc per mapper (converted BAMs only)" )
} else {
  p2 = tab %>% ggplot(aes(x=mapper, y=m_ntc, col=mapper, shape=true_isoform, group=paste0(mapper, true_isoform, classification))) + 
    geom_point() +
    facet_wrap(classification~.) +
    ggtitle( "Mean n_tc per mapper (converted BAMs only)" ) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}

tab = reads %>% filter(is_converted_bam==1) %>% 
  group_by(classification, mapper, condition_id, true_isoform) %>% summarise(m_seqerr_perc=mean(n_true_seqerr, na.rm=T))
if (params$plot_type=='by_condition') {
  p3 = tab %>% ggplot(aes(x=condition_id, y=m_seqerr_perc, col=mapper, linetype=true_isoform, group=paste0(mapper, true_isoform, classification))) + 
    geom_line() + geom_point() +
    facet_wrap(classification~.) +
    ggtitle( "Mean seq error per mapper (converted BAMs only)" )
} else {
  p3 = tab %>% ggplot(aes(x=mapper, y=m_seqerr_perc, col=mapper, shape=true_isoform, group=paste0(mapper, true_isoform, classification))) + 
    geom_point() +
    facet_wrap(classification~.) +
    ggtitle( "Mean seq error per mapper (converted BAMs only)" ) + 
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}

plot_grid(p1,p2,p3) 
ggsave(paste0(params$out_dir,'/read_performance.pdf'), width = 12, height = 8)

```


## per-gene stats
```{r}
gn2tid = function(gn) {
  return(unique( d[['tid_performance']] %>% filter(gene_name == gn) %>% pull(tid) ))
}

plot_gene = function(gname) {
  tid=gn2tid(gname)
  if (params$plot_type=='by_condition') {
    d[['tid_performance']] %>% filter(tid==!!gn2tid(gname)) %>% ggplot(aes(x=condition_id, y=F1, col=mapper, linetype=iso, group=paste0(mapper, iso))) + 
      geom_line() + geom_point() +
      ggtitle( paste0("F1 for ", gname) )
  } else {
  d[['tid_performance']] %>% filter(tid==!!gn2tid(gname)) %>% ggplot(aes(x=mapper, y=F1, col=mapper, shape=iso, group=paste0(mapper, iso ))) + 
    geom_line() + geom_point() +
    ggtitle( paste0("F1 for ", gname) ) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  }
}

p1=plot_gene('Actb')
p2=plot_gene('Mat2a')
p3=plot_gene('Gm10243')
p4=plot_gene('Ybx1-ps2')
p5=plot_gene('Gm23306')
p6=plot_gene('Gm38323')
my_plot_grid(list(p1,p2,p3,p4,p5,p6), main = "example genes")
ggsave(paste0(params$out_dir,'/example_genes.pdf'), width = 10, height = 8)
```     


## splice site mappability
```{r}
plot_mappability_by_bin = function(tab, type='donor', subtitle='', facet=NA) {
  group_vars=c('mapper', 'condition_id')
  if (!is.na(facet)) {
    # rename by addint '(n=xxx)')
    cnts = tab %>% 
      group_by_at(facet) %>% 
      summarise(n_tid=n_distinct(tid)) %>% 
      setNames(., c("a", "b")) %>% mutate(c=paste0(a,' (n=',b,')'))
    labs = setNames(cnts$c, cnts$a)
    tab[['facet']]=factor( labs[tab[[facet]]], levels=cnts$c)
    group_vars=c(group_vars, 'facet')
  }
  if (type=='donor') {
    dat = calc_ci(tab %>% group_by_at(group_vars), mean_don_prec, min_lower_ci=0) %>% mutate(name='mean_don_prec')
  } else {
    dat = calc_ci(tab %>% group_by_at(group_vars), mean_acc_prec, min_lower_ci=0) %>% mutate(name='mean_acc_prec')
  }
  if ( params$plot_type=='by_condition') {
    p=dat %>% 
      ggplot(aes(x=condition_id, y=col.mean, col=mapper, group=mapper)) + 
        geom_ribbon(aes(ymin=lower.ci, ymax=upper.ci), linetype=2, alpha=0.2) +
        geom_line() +
        ggtitle(paste0(type, " splice site precision  (n=",length(unique(tab$tid)),")"), subtitle)
    if (!is.na(facet)) {
      p=p+facet_wrap(facet~.)
    }
  } else {
    
  }
  return(p)
}

tab = d[['splice_site_mappability']] %>% mutate(
    mapper=factor(mapper),
    condition_id=factor(condition),
    don_prec=ifelse(don_TP_reads+don_FP_reads>0,don_TP_reads/(don_TP_reads+don_FP_reads),0), 
    acc_prec=ifelse(acc_TP_reads+acc_FP_reads>0,acc_TP_reads/(acc_TP_reads+acc_FP_reads),0)
    ) %>% 
  group_by(mapper, condition_id, tid) %>% 
  summarise(mean_don_prec=mean(don_prec), 
            mean_acc_prec=mean(acc_prec),
            mean_don_win_min_map=mean(don_win_min_map),
            mean_acc_win_min_map=mean(acc_win_min_map)
            ) %>% 
  mutate(don_win_map_cat=case_when(
            mean_don_win_min_map<0.25 ~ 'low',
            mean_don_win_min_map>0.75 ~ 'high',
            T ~ 'medium'),
         acc_win_map_cat=case_when(
            mean_acc_win_min_map<0.25 ~ 'low',
            mean_acc_win_min_map>0.75 ~ 'high',
            T ~ 'medium')
    ) %>% 
  mutate(don_win_map_cat=factor(don_win_map_cat, levels=c('high', 'medium', 'low')),
         acc_win_map_cat=factor(acc_win_map_cat, levels=c('high', 'medium', 'low'))
         )

p1 = plot_mappability_by_bin(tab, type='donor')
p2 = plot_mappability_by_bin(tab, type='acceptor')
p3 = plot_mappability_by_bin(tab, type='donor', subtitle='by SJ mappability', facet='don_win_map_cat')
p4 = plot_mappability_by_bin(tab, type='acceptor', subtitle='by SJ mappability', facet='acc_win_map_cat')
my_plot_grid(list(p1,p2,p3,p4), main = "Splice site mappability")
ggsave(paste0(params$out_dir,'/splice_site_mappability.pdf'), width = 12, height = 8)

# this looks a bit weird: why is STAR mappability so good for SJ with low SJ mappability?
# check: tab %>% group_by(mapper,  don_win_map_cat) %>% summarise(m=mean(mean_don_prec)) 
```

## splice-site mappability corrections

Can we correct for low SJ mappability? 
- find SJ with low mappabililty across conversion rates + mappers
- filter: remove genes with too-short exons (e.g., chr13:55,132,787-55,133,731 or chr6:115,227,343-115,259,294) where there are no simulated spliced reads!


```{r map_corr}
ss_perf_per_intron = d[['splice_site_performance']] %>% 
  group_by(mapper,condition_id,tid,rnk,mappability_mean) %>% 
  mutate(true_mat=spl_FN+spl_TP,
         found_mat=spl_FP+spl_TP,
         true_pre=don_FN+don_TP+acc_FN+acc_TP,
         found_pre=don_FP+don_TP+acc_FP+acc_TP,
         ) %>% 
  mutate(
         true_spliced=ifelse(true_pre>0,true_mat/(true_pre+true_mat), NA),
         found_spliced=ifelse(found_pre>0,found_mat/(found_pre+found_mat), NA),
         diff=abs(true_spliced-found_spliced),
         low_map=diff>0.01
  ) %>% ungroup()

CUTOFF=0.05
max_diff_introns=ss_perf_per_intron %>% 
  filter(mapper != 'truth') %>% 
  group_by(intron_id) %>% 
  summarise(max_diff=max(diff, na.rm = T)) %>% 
  mutate(low_map=max_diff>!!CUTOFF)
max_diff_introns %>% dplyr::count(low_map) %>% ggplot(aes(x=low_map, y=n)) + geom_bar(stat='identity') + ggtitle("number of low_mappability introns")


high_map_introns = max_diff_introns %>% filter(!low_map) %>% pull(intron_id)
  
ss_perf_per_tid = d[['splice_site_performance']] %>% 
  group_by(mapper,condition_id,tid,rnk,mappability_mean) %>% 
  summarise(true_mat=sum(spl_FN)+sum(spl_TP),
         found_mat=sum(spl_FP)+sum(spl_TP),
         true_pre=sum(don_FN)+sum(don_TP)+sum(acc_FN)+sum(acc_TP),
         found_pre=sum(don_FP)+sum(don_TP)+sum(acc_FP)+sum(acc_TP),
         .groups = 'drop'
         ) %>% 
  mutate(
         true_spliced=ifelse(true_pre>0,true_mat/(true_pre+true_mat), NA),
         found_spliced=ifelse(found_pre>0,found_mat/(found_pre+found_mat), NA),
  ) %>%  
  mutate(num_exons=case_when(rnk<=5 ~ as.character(rnk), 
                             rnk>5 ~ '>5',
                             T ~ 'NA'),
         mappability=case_when(mappability_mean<0.2 ~ 'low',
                               mappability_mean>0.9 ~ 'high',
                               T ~ 'medium') 
  )  %>% 
  mutate(num_exons=factor(num_exons, levels=c('1','2','3','4','5','>5')),
         mappability=factor(mappability, levels=c('high', 'medium', 'low'))
  ) %>% 
  merge(d[['transcript_info']], by='tid', all.x=T)


ss_perf_per_tid_filtered = d[['splice_site_performance']] %>%
  filter(intron_id %in% high_map_introns ) %>% 
  group_by(mapper,condition_id,tid,rnk,mappability_mean) %>% 
  summarise(true_mat=sum(spl_FN)+sum(spl_TP),
         found_mat=sum(spl_FP)+sum(spl_TP),
         true_pre=sum(don_FN)+sum(don_TP)+sum(acc_FN)+sum(acc_TP),
         found_pre=sum(don_FP)+sum(don_TP)+sum(acc_FP)+sum(acc_TP),
         .groups = 'drop'
         ) %>% 
  mutate(
         true_spliced=ifelse(true_pre>0,true_mat/(true_pre+true_mat), NA),
         found_spliced=ifelse(found_pre>0,found_mat/(found_pre+found_mat), NA),
  ) %>%  
  mutate(num_exons=case_when(rnk<=5 ~ as.character(rnk), 
                             rnk>5 ~ '>5',
                             T ~ 'NA'),
         mappability=case_when(mappability_mean<0.2 ~ 'low',
                               mappability_mean>0.9 ~ 'high',
                               T ~ 'medium') 
  )  %>% 
  mutate(num_exons=factor(num_exons, levels=c('1','2','3','4','5','>5')),
         mappability=factor(mappability, levels=c('high', 'medium', 'low'))
  ) %>% 
  merge(d[['transcript_info']], by='tid', all.x=T)


# p1=ss_perf_per_tid %>% 
#   ggplot(aes(x=condition_id, y=found_spliced)) +
#     geom_boxplot() +
#     facet_wrap(mapper~.) +
#     geom_hline(yintercept = 1/3, linetype='dotted', color='grey') 
# p2=ss_perf_per_tid %>% 
#   filter(rnk==2) %>% 
#   ggplot(aes(x=condition_id, y=found_spliced)) +
#     geom_boxplot() +
#     facet_wrap(mapper~.) +
#     geom_hline(yintercept = 1/3, linetype='dotted', color='grey') 

# unfiltered
p1=calc_ci(ss_perf_per_tid %>% group_by(mapper, condition_id, num_exons, mappability), found_spliced, min_lower_ci=0) %>% 
  ggplot(aes(x=condition_id, y=col.mean, col=mapper, group=mapper)) +
    geom_ribbon(aes(ymin=lower.ci, ymax=upper.ci), linetype=2, alpha=0.2) +
    geom_line() +
    ggtitle("frac_mature (uncorrected)", 
            paste0("n_tids=", ss_perf_per_tid %>% summarise(n=n_distinct(tid)))) +
    facet_grid(mappability~num_exons) +
    geom_hline(yintercept = 1/3, linetype='dotted', color='grey') 

p2=calc_ci(ss_perf_per_tid %>% filter(n_overlapping<1) %>% group_by(mapper, condition_id, num_exons, mappability), found_spliced, min_lower_ci=0) %>% 
  ggplot(aes(x=condition_id, y=col.mean, col=mapper, group=mapper)) +
    geom_ribbon(aes(ymin=lower.ci, ymax=upper.ci), linetype=2, alpha=0.2) +
    geom_line() +
    ggtitle("frac_mature (uncorrected, no overlapping)", 
            paste0("n_tids=", ss_perf_per_tid %>% filter(n_overlapping<1) %>% summarise(n=n_distinct(tid)))) +
    facet_grid(mappability~num_exons) +
    geom_hline(yintercept = 1/3, linetype='dotted', color='grey') 

# filtered 
p3=calc_ci(ss_perf_per_tid_filtered %>% group_by(mapper, condition_id, num_exons, mappability), found_spliced, min_lower_ci=0) %>% 
  ggplot(aes(x=condition_id, y=col.mean, col=mapper, group=mapper)) +
    geom_ribbon(aes(ymin=lower.ci, ymax=upper.ci), linetype=2, alpha=0.2) +
    geom_line() +
    ggtitle("frac_mature (corrected)", 
            paste0("n_tids=", ss_perf_per_tid_filtered %>% summarise(n=n_distinct(tid)))) +
    facet_grid(mappability~num_exons) +
    geom_hline(yintercept = 1/3, linetype='dotted', color='grey') 

p4=calc_ci(ss_perf_per_tid_filtered %>% filter(n_overlapping<1) %>% group_by(mapper, condition_id, num_exons, mappability), found_spliced, min_lower_ci=0) %>% 
  ggplot(aes(x=condition_id, y=col.mean, col=mapper, group=mapper)) +
    geom_ribbon(aes(ymin=lower.ci, ymax=upper.ci), linetype=2, alpha=0.2) +
    geom_line() +
    ggtitle("frac_mature (corrected, no overlapping)", 
            paste0("n_tids=", ss_perf_per_tid_filtered %>% filter(n_overlapping<1) %>% summarise(n=n_distinct(tid)))) +
    facet_grid(mappability~num_exons) +
    geom_hline(yintercept = 1/3, linetype='dotted', color='grey') 
my_plot_grid(list(p1,p2,p3,p4), main = "Splice site analysis")

```


## todo, example genes

- Ttn
- Npm1
- Gas5 (snoRNA host gene)

## Genes with big mapper differences
```{r big_diff}
tab = d[['tid_performance']] %>% 
  filter(mapper!='truth') %>% 
  select(condition_id, gene_name, tid, iso, chromosome,start,end,  mapper, F1) %>% 
  pivot_wider(names_from = mapper, values_from = F1, id_cols = c(condition_id, gene_name, tid, iso, chromosome,start,end)) %>% 
  mutate(f1_diff=STAR-HISAT3N) 
tab %>% ggplot(aes(x=condition_id, y=f1_diff, fill=iso)) +
  geom_boxplot() +
  geom_hline(yintercept = 0, col='grey', linetype='dotted') +
  ggtitle("Difference in F1 value per transcript") 

# example gene where there are no mapped reads in HISAT3N / cr==0 for some reason
d[['tid_performance']] %>% filter(tid=='ENSMUST00000179058.1') %>% select(tid, gene_name,TP,FP,FN,F1, everything())

```

