---
title: "splice_sim_celltype_mappability.Rmd"
author: "niko.popitsch@imba.oeaw.ac.at"
documentclass: article
fontsize: 10pt
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 14
    fig_height: 10
    fig_caption: true
    df_print: paged
  pdf_document:
    fig_width: 12
    fig_height: 10
    fig_caption: true
params:
    gff_file:
       value: "/groups/ameres/Niko/ref/genomes/mm10/annotation/gencode.vM21.annotation.nochr.sorted.gff3.gz"
    expression_data:
       value: "/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/splice_sim_paper/expression_data/E-MTAB-6798-query-results.tpms.tsv"
    mesc_data:
       value: "/groups/ameres/Niko/projects/Ameres/splicing/data/slamstr/pooled.mm10.all/pooled.mm10.all.Slamstr.stats.trans.final.tsv.gz"
    salmon_data:
      value: "/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/splice_sim_paper/expression_data"
    slamstr_result:
      value: "/groups/ameres/Niko/projects/Ameres/splicing/data/slamstr/pooled.mm10.all/pooled.mm10.all.final.rds"
    data_dir:
       value: "/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/splice_sim_paper/replicate_data_kss/"
    out_dir:
       value: "/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/splice_sim_paper/analysis_kss/"
---
<style type="text/css">
body, td {
   font-size: 10px;
}
code.r{
  font-size: 10px;
}
pre {
  font-size: 10px
}
div.blue pre { background-color:lightblue; }
div.blue pre.r { background-color:blue; }
</style>

<!--
rmarkdown::render('splice_sim_paper.Rmd', 'html_document')
-->

# INIT

```{r setup, include=FALSE}
require(GenomicFeatures)
require(tximport)
require(data.table)
require(goseq)
require(AnnotationDbi)
require(BiocGenerics)
require(parallel)
require(tidyr)
require(dplyr)
require(dtplyr)
require(ggplot2)
require(scales)
require(rjson)
require(stringr)
require(VGAM)
require(cowplot)
require(arrow)
require(tictoc)
require(ggpubr)
require(minpack.lm)
require(readr)
require(testthat)
require(RColorBrewer)
require(skimr)
require(ggforce)
require(writexl)
require(xfun)
require(forcats)
require(glue)

rename=dplyr::rename
select=dplyr::select
arrange=dplyr::arrange

# NB install arrow with snappy on Rstudio server
# Sys.setenv(ARROW_WITH_SNAPPY = "ON")
# Sys.setenv(NOT_CRAN="true")
# Sys.setenv(LIBARROW_BINARY="FALSE")
# install.packages("arrow", repos = "https://arrow-r-nightly.s3.amazonaws.com")

# to ensure stripping '\0' (nul) from character vector
options(arrow.skip_nul = TRUE)

# global theme
ggplot2::theme_set(theme_light())

# enable tidylog (has overhead!)
#require(tidylog)
options("tidylog.display" = list()) # turn tidylog off
options(scipen=10000) # to avoid scientific notation y axis 

# load resources/functions
source('splice_sim_resources.R')
```


# data

```{r data, include=F, echo=F, cache=T, eval=F}
m = read_rds(paste0(params$data_dir, '/meta.rds'))


# RNAseq counts / TPM values for regular rnaseq
reg_rnaseq_exp = load_table(params$expression_data) %>% as_tibble()
celltypes = names(reg_rnaseq_exp)[! names(reg_rnaseq_exp) %in% c('Gene ID', 'Gene Name')]
top_genes_rnaseq=tibble()
for ( ct in celltypes ) {
  top_genes_rnaseq = top_genes_rnaseq %>% 
    bind_rows(
      reg_rnaseq_exp %>% 
        select(gene_id=`Gene ID`, gene_name=`Gene Name`, tpm=!!rlang::sym(ct)) %>% 
        left_join(m[['ga']], by='gene_name') %>% 
        filter(!is.na(tpm) & !is.na(tid)) %>% 
        filter(tpm>10) %>% 
        arrange(desc(tpm)) %>% 
        mutate(ds=ct, cat='RNA-seq')      
    )
}

# load mesc tids
# these are the big3 genes. They are from
# load_table('/groups/ameres/Niko/projects/Ameres/splicing/data/slamstr/pooled.mm10.all/pooled.mm10.all.Slamstr.stats.trans.final.tsv.gz') %>% filter(TAB_ALL_tpm_180min>10) %>% as_tibble() %>% nrow()
# this is a pulse TT-SLAMseq dataset (2 replicates pooled together).
# top_genes_mESC = load_table(params$mesc_data) %>% as_tibble() %>% 
#         select(tid=transcript_id, TAB_ALL_tpm_180min) %>% 
#         filter(TAB_ALL_tpm_180min>10) %>% 
#         left_join(m[['ga']], by='tid') %>% 
#         mutate(ds='mESC', cat='TT-Slamseq')


# add all annotated mouse genes
top_genes_all = m[['ga']] %>% 
        mutate(ds='mm10', cat='all_annotated')


# add TPM estimates from salmon for r12 input + ip
# create tx2gene
txdb <-makeTxDbFromGFF(params$gff_file, format="gff")
k <- keys(txdb, keytype = "GENEID")
df <- AnnotationDbi::select(txdb, keys = k,  columns = "TXNAME", keytype = "GENEID")
tx2gene <- df[, 2:1]
head(tx2gene)
# read salmon data
files=c(
  paste0(params$salmon_data, '/salmon_r12_in_180/quant.sf'),
  paste0(params$salmon_data, '/salmon_r12_ip_180/quant.sf')
)
names(files)=c('in', 'ip')
txi=tximport(files,type="salmon", tx2gene=tx2gene, countsFromAbundance="scaledTPM")
top_genes_salmon=m[['tx']] %>% select(tid) %>% 
  left_join( txi$abundance %>% 
               as_tibble(rownames = 'gene_id') %>% 
               left_join(tx2gene, by=c('gene_id'='GENEID')) %>% 
               rename(tid=TXNAME), by='tid') %>% 
  filter(`in`>10) %>% 
  left_join(m[['ga']], by='tid') %>% 
  mutate(ds='mESC', cat='RiboZero')

# write gene set 
# top_genes_salmon %>% 
#   select(transcript_id=tid) %>% 
#   write_tsv('/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/testruns/ref/mESC_expressed_tx_ribozero.tsv')

# write bed file
# top_genes_salmon %>% 
#   left_join(m[['tx']], by='tid') %>% 
#   select(chromosome, start, end, gene_name) %>% 
#   arrange(chromosome,start) %>% 
#   write_tsv('/groups/ameres/Niko/projects/Ameres/splicing/splice_sim/splice_sim_paper/expression_data/mESC_expressed_tx_ribozero.bed', col_names=FALSE)


# merge
top_genes = bind_rows(
  top_genes_all,
  #top_genes_mESC,
  top_genes_salmon,
  top_genes_rnaseq
) 
top_genes = top_genes %>% 
  left_join(m[['tx']], by='tid') %>% 
  mutate(ds=factor(ds, levels=unique(top_genes$ds )),
         cat=factor(cat, levels=unique(top_genes$cat))) %>% 
  filter(!is.na(chromosome))
  

# top_genes %>% 
#   group_by(cat, ds, mappability) %>% 
#   count() %>% 
#   ggplot(aes(ds, n, fill=mappability)) +
#   geom_col() +
#   coord_flip() +
#   my_scales() +
#   #facet_wrap(cat~., scales = 'free_x') +
#   ggtitle("Mappability category distributions for expressed/transcribed genes")

top_genes %>% group_by(cat, ds, mappability) %>% count() %>% 
  group_by(cat, ds) %>% mutate(s=sum(n), frac=n/s) %>% 
  group_by(cat, mappability) %>% 
  summarise(frac=mean(frac)) %>% 
  ggplot(aes(cat, frac, fill=mappability)) +
  geom_col() +
  coord_flip() +
  my_scales() +
  #facet_wrap(cat~., scales = 'free_x') +
  ggtitle("Mappability category distributions for expressed/transcribed genes") +
  ylab("Fraction of genes") + xlab("")



ggsave(paste0(params$out_dir,'/qc_map_cat_dist.pdf'), width = 10, height = 6)
#+scale_y_log10()

# Distribution of mappability categories for expressed genes (TPM>10) in (i) regular RNAseq data (TPM value tables downloaded from EBI ExpressionAtlas project E-MTAB-6798; plot shows mean category fractions over 94 datasets), (ii) Ribo-zero RNAseq data from mESC cells, (iii) all transcripts used in this study. Ribo-zero data contains a considerably higher fraction of low and medium mappability transcripts (>50%)


```

